test/suite/wttest.py

```diff
@@ -284,6 +307,40 @@ def skipTest(self, reason):
         self.skipped = True
         super(WiredTigerTestCase, self).skipTest(reason)
 
+    # Override unittest.run.
+    # We do this so we can intercept calls to the test method.
+    def run(self, result=None):
+        self._savedTestMethodName = self._testMethodName
+        self._testMethodName = "runTestMethod"
+        try:
+            super(WiredTigerTestCase, self).run(result)
+        finally:
+            self._testMethodName = self._savedTestMethodName
+
+    # This function is called in advance of every test method
+    # invocation.  It retries the test method whenever
+    # it fails due to a rollback error, up to a point.
+    def runTestMethod(self):
+        rollbacksAllowed = self._rollbacksAllowedPerTest
+        finished = False
+        method = getattr(self, self._savedTestMethodName)
+        WiredTigerTestCase._testsRun += 1
+        while not finished and rollbacksAllowed > 0:
+            try:
+                method()
+                finished = True
+            except wiredtiger.WiredTigerRollbackError:
+                WiredTigerTestCase._retriesAfterRollback += 1
+                self.prexception(sys.exc_info())
+                if rollbacksAllowed <= 0:
```

Comment:
Minor: I think `rollbacksAllowed` can never be < 0 here because of the while loop condition. Can/should you check for == 0? 

Also: Do you want to consider counting how many rollbacks per test and comparing at the end? That way if a long test with a lot of transactions triggers this, we get signal on considering what value the `_rollbacksAllowedPerTest` global setting should be.

Reply:
I'll fix the check.

> Do you want to consider counting how many rollbacks per test and comparing at the end?

@sueloverso I'm not grasping what you are suggesting here.  _rollbacksAllowedPerTest is for each individual test.  On a rollback error, the entire test is restarted, with a new connection etc.  So it kind of doesn't matter how many transactions there are in a test (and we don't have an easy way to know that), if one fails with a rollback, we retry the test.

I see the code is not explicitly closing the old connection on a retry, so it would be orphaned.  I'll fix that - I think I need to clear out the directory also.

Reply:
Ah - thanks for the explanation. I forgot that it was retrying the entire test not per transaction.